pandas/core/internals/blocks.py

```diff
@@ -1005,11 +1014,21 @@ def putmask(self, mask, new) -> list[Block]:
         new = extract_array(new, extract_numpy=True)
 
         if noop:
+            if using_cow:
+                return [self.copy(deep=False)]
             return [self]
 
         try:
             casted = np_can_hold_element(values.dtype, new)
+
+            if using_cow and self.refs.has_reference():
+                # Do this here to avoid copying twice
+                values = values.copy()
+                self = type(self)(values, self._mgr_locs, ndim=self.ndim)
+
             putmask_without_repeat(values.T, mask, casted)
+            if using_cow:
+                return [self.copy(deep=False)]
```

Comment:
We don't need to shallow copy if we already did a hard copy above?

Reply:
Ah, but the intermediate block will go out of scope, so making a shallow copy here will still only have a single ref in that case?

Reply:
Exactly, it does not hurt, that’s why I wanted to avoid making the check more complex